### ISSUE TITLE
Allow for map type variables

### ISSUE THREAD

**Issue by stevenroose** [stevenroose/gonfig#1]
A map type would ultimately allow for setting any suboption to that parent option.

```go
var config struct{
    Var map[string]interface{}
}{}
```

This could be set like this: `--var.somekey somevalue --var.otherkey othervalue` or envs `PREFIX_VAR_SOMEKEY`.

Would not be that hard to implement, but I currently don't have a need for this.

**Comment by robertgzr** [comment]
This would definitely be a nice feature to have but I would like for this to not break TOML quoted keys like viper does (https://github.com/spf13/viper/pull/399)

So if there is a way to access a map it would be nice to provide a way to change that delimiter.

**Comment by stevenroose** [comment]
@robertgzr Can you elaborate on that a bit? How would it break TOML and what are _quoted keys_?

I just imagined that an option named "extraopts" defined as a `map[string]interface{}` in TOML would be:
```toml
[extraopts]
key1 = "val1"
key2 = 2
```
Where you can use any keys and access them with `config.ExtraOpts["key1"]`. 

With flags this would be `--extraopts.key1 value1` and with env `PREF_EXTRAOPTS_KEY1=value1`.

**Comment by robertgzr** [comment]
@stevenroose 
My use case would be that the config file has a fields to specify mappings of one file to another:
```toml
[copythis]
  "/path/to/destination.txt" = "/path/to/source.txt"
```
or more examples in the [TOML spec](https://github.com/toml-lang/toml/blob/master/versions/en/toml-v0.4.0.md#table)

But thinking about this you could have the same problem using JSON:
```json
{
  "copythis": {
    "/path/to/destination.txt": "/path/to/source.txt"
  }
}
```

viper [splits](https://github.com/spf13/viper/blob/master/viper.go#L868) the keys when it accesses a map in this way.

But it looks like with the way `gonfig` is implemented that's not the issue here. You would just have to keep in mind that the map key might not be a valid identifier with JSON/TOML.
Maybe you could run a [strings.Replacer](https://godoc.org/strings#Replacer) over the key before using it as a flag or env or allow to exclude fields to be accessed from flag/env?

**Comment by stevenroose** [comment]
I might be using this in a project of mine soon, so I will try to implement this. I'm not entirely sure what your concern with TOML quoted keys is, but I'll ping you when there is a PR so that you can verify if it causes you trouble.

**Comment by stevenroose** [comment]
@robertgzr I forgot about this issue. I implemented map types here in 2d1936f9adb3a951d62c2dd0c097d442bd39dfa5, could you take a look and see if you spot any problems?
